### PR TITLE
feat: Add token CREATE route

### PR DESCRIPTION
**Pull request by ian-fox**
* `POST`ing to /tokens will create a new token for the logged in user.

### CODE REVIEW

**Comment by ian-fox** [comment]
For some reason the tests are all passing on my machine but failing on screwdriver, investigating now.

**Comment by minz1027** [comment]
@ian-fox Ian, the last merge also failed the same unit test, probably the same reason.

**Comment by ian-fox** [comment]
@minz1027 Seems to be something to do with dependencies, got the tests to start failing on my machine after I `rm`'d the node_modules folder and did a fresh `npm install`

File: test/plugins/tokens.test.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+const getUserMock = (user) => {
+    const mock = hoek.clone(user);
+
+    mock.tokens = sinon.stub();
+
+    return mock;
+};
+
+const tokensGetterMock = tokens =>
+    new Promise(resolve => resolve(tokens));
```
Comment (minz1027) [inline test/plugins/tokens.test.js]: This can be tokensGetterMock = tokens => Promise.resolve(tokens);

File: test/plugins/tokens.test.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+const getUserMock = (user) => {
+    const mock = hoek.clone(user);
+
+    mock.tokens = sinon.stub();
+
+    return mock;
+};
+
+const tokensGetterMock = tokens =>
```
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: `const tokensGetterMock = tokens => Promise.resolve(tokens)`

File: test/plugins/tokens.test.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            port: 1234
+        });
+
+        server.auth.scheme('custom', () => ({
+            authenticate: (request, reply) => reply.continue({})
+        }));
+        server.auth.strategy('token', 'custom');
+        server.auth.strategy('session', 'custom');
+
+        server.register([{
```
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: `server.register` can also return a promise.

File: test/plugins/tokens.test.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        userMock = getUserMock({
+            username,
+            id: userId
+        });
+        userMock.tokens = tokensGetterMock([]);
+        userFactoryMock.get.withArgs({ username }).resolves(userMock);
+    });
+
+    afterEach(() => {
+        server = null;
```
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: Probably want to call `server.stop` before removing the only reference to it

File: test/plugins/tokens.test.js (hunk 3 above)
Comment (ian-fox) [inline test/plugins/tokens.test.js]: So I should do the mock stuff first, and then `return` server.register?

File: plugins/tokens/create.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            }
+        },
+        handler: (request, reply) => {
+            const tokenFactory = request.server.app.tokenFactory;
+            const userFactory = request.server.app.userFactory;
+            const username = request.auth.credentials.username;
+            let user;
+
+            return userFactory.get({ username })
+            .then((theUser) => {
```
Comment (d2lam) [inline plugins/tokens/create.js]: Hmm why not just leave this as `.then((user) => ...`, and have the `.then(tokens)` inside this user scope? I think that would be cleaner.

File: plugins/tokens/create.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                user = theUser;
+
+                return user.tokens;
+            })
+            .then((tokens) => {
+                // Make sure the name is unique
+                const match = tokens && tokens.find(t => t.name === request.payload.name);
+
+                if (match) {
+                    throw boom.conflict('Token with that name already exists');
```
Comment (d2lam) [inline plugins/tokens/create.js]: Would be clearer to have the `${name}` here

File: test/plugins/tokens.test.js [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+const tokensGetterMock = tokens => Promise.resolve(tokens);
+
+const generateProfile = sinon.stub();
+
+describe('token plugin test', () => {
+    const username = 'ifox';
+    const userId = testToken.userId;
+    let tokenFactoryMock;
+    let userFactoryMock;
+    let authMock;
```
Comment (minz1027) [inline test/plugins/tokens.test.js]: why we need the authMock here? This is only creating a token and save it in datastore. We shouldn't need to mock auth stuff.

File: test/plugins/tokens.test.js (hunk 3 above)
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: Yup, that approach is valid

File: test/plugins/tokens.test.js (hunk 7 above)
Comment (ian-fox) [inline test/plugins/tokens.test.js]: Right, remnant from something else.

**Pull request merged**